src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,526 @@
+                var builder = ArrayBuilder<SyntaxNode>.GetInstance(names.Length);
+                for (var i = 0; i < names.Length; ++i)
+                {
+                    builder.Add(CreateImport(generator, names[i], withFormatterAnnotation));
+                }
+
+                return builder.ToImmutableAndFree();
+            }
+
+            private SyntaxNode CreateImport(SyntaxGenerator syntaxGenerator, string name, bool withFormatterAnnotation)
+            {
+                var import = syntaxGenerator.NamespaceImportDeclaration(name);
+                if (withFormatterAnnotation)
+                {
+                    import = import.WithAdditionalAnnotations(Formatter.Annotation);
+                }
+                return import;
+            }
+
+            private async Task<(Solution, ImmutableArray<DocumentId>)> ChangeNamespaceToMatchFoldersAsync(Solution solution, DocumentId id, CancellationToken cancellationToken)
+            {
+                var document = solution.GetDocument(id);
+                var targetNamespace = _state.TargetNamespace;
+
+                var declarationRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var container = declarationRoot.DescendantNodes().FirstOrDefault(node => node is TNamespaceDeclarationSyntax) ?? declarationRoot;
+
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var declaredSymbols = GetDeclaredSymbolsInContainer(semanticModel, container, cancellationToken);
+
+                var documentSet = ImmutableHashSet.Create(document);
+                var editor = await DocumentEditor.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+
+                var refLocationsInCurrentDocument = new List<ReferenceLocation>();
+                var refLocationsInOtherDocuments = new List<ReferenceLocation>();
+
+                foreach (var declaredSymbol in declaredSymbols)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    var refSymbols = await SymbolFinder.FindReferencesAsync(
+                      declaredSymbol,
+                      solution,
+                      cancellationToken).ConfigureAwait(false);
+
+                    var refLocationsForSymbol = refSymbols.Where(refSymbol => refSymbol.Definition == declaredSymbol)
+                        .SelectMany(refSymbol => refSymbol.Locations);
+
+                    foreach (var refLocation in refLocationsForSymbol)
+                    {
+                        if (refLocation.Document.Id == document.Id)
+                        {
+                            refLocationsInCurrentDocument.Add(refLocation);
+                        }
+                        else
+                        {
+                            Debug.Assert(!PathUtilities.PathsEqual(refLocation.Document.FilePath, document.FilePath));
```

Comment:
i don't see how that's safe to assert given things like linked files... can you clarify?

Reply:
Because how to determine whether to trigger the refactoring, the only file that can be linked to `document` is its counterpart in a project targeting another framework in a multi-targeting project, which can't have reference to anything defined in `document`.